doc/sphinxman/source/add_tests.rst

```diff
@@ -30,108 +30,312 @@
-template job or that test error handling or that are PsiAPI rather than
-PSIthon focused. In these cases, you'll want to add to the second test
-suite that lives at :source:`tests/pytests`. Presently, the "normal"
-(everything in the ``tests/`` directory that isn't in ``tests/pytests/``)
-are run through :program:`ctest`, while the pytests are run through :program:`pytest`. In
-future, all will be run through Pytest, but the former will still be
-run as PSIthon (``psi4 input.dat``) while the latter will still be
-run as PsiAPI (``import psi4``). In other words, in designing a test,
-choose its mode based on whether PSIthon or PsiAPI suits it better and
-whether it's a simple model for users (probably PSIthon) or for expert
-users (probably PsiAPI). Both will continue to work in future.
-
-In developing a Pytest test, you probably want to edit it in place,
-rather than running :program:`make` after each change. Easiest is from
-<objdir>, run ``pytest ../tests/pytests``. Add any filters (``-k
-test_name_fragment``) or parallelism (``-n <N>`` if ``pytest-xdist``
-installed) or print test names (``-v``) or print warnings (``-rws``). To
-see stdout output from an otherwise passing test, easiest to add ``assert
-0`` at its end to trigger failure. An important point is that because
-they're PsiAPI, ``import psi4`` is happening, so the <objdir> |PSIfour|
-module must be in :envvar:`PYTHONPATH`. Also, any call to QCEngine is
-using ``which psi4``, so the <objdir> |PSIfour| executable must be in
-:envvar:`PATH`. The easiest way to prepare your local environment is to
-execute the printout of ``<objdir>/stage/bin/psi4 --psiapi``.
+To create a new test case, either create a new file or add to an existing file under :source:`tests/pytests`.
 
 * Test must be in the :source:`tests/pytests/` directory.
 * Test file name must start with ``test_``. This is how pytest knows to collect it.
-* Test file may contain many tests. To be recognized as a test, the Python function must start with ``test_``.
+* Test file may contain many tests, each of which is an ordinary Python function with name starting ``test_``. This is how pytest knows to collect it.
 * No registration required to bring a test to pytest's attention.
+* No registration required to bring a test to CMake's attention. If a test needs additional files, register them in :source:`psi4/CMakeLists.txt`.
+
+A few notes on test contents:
+
+* Import testing functions from ``utils`` and use Python assert: ``assert compare_values(expected, computed, ...)``.
+* Don't worry about cleaning up files or resetting options. There's a function in :source:`tests/pytests/conftest.py` that does so between every test.
+* Especially if using data or functions from outside a test, run a variety of tests at different parallelisms. If tests fail that pass when run alone, you've got a function of the same name changing state or some similar correctable phenomenon.
```

Comment:
What is "parallelism"? Based on the example you give, I don't think you mean "number of threads."